larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/ShowerESTAREnergy_tool.cc

```diff
@@ -0,0 +1,293 @@
+    // Get shower centre 
+    TVector3 showercentre = IShowerTool::GetLArPandoraShowerAlg().ShowerCentre(spacepoints);
+    
+    // Get the charge weighted shower centre - sum(charge*position)/sum(charge)
+    // Charge is obtained from the lifetime corrected hits and hits are obtained from the spacepoints 
+    TVector3 chargecentre = IShowerTool::GetLArPandoraShowerAlg().ShowerCentre(clockData, detProp, spacepoints, fmhsp);
+    
+    // Get space charge corrected E-field
+    if(fSCECorrectEField){
+      // Check the shower centre exists and is sensible
+      // Think the charge weighted centre can take nan values which causes the E-field calculation to fail
+      if(showercentre.Mag() >= 0 && chargecentre.Mag() >=0){ 
+        localEfield = IShowerTool::GetLArPandoraShowerAlg().SCECorrectEField(nominal_Efield, showercentre);
+        localEfield_cweighted = IShowerTool::GetLArPandoraShowerAlg().SCECorrectEField(nominal_Efield, chargecentre);
+      }
+      else{
+        mf::LogWarning("ShowerESTAREnergy") << "Shower centre calculation doesn't look to be sensible. Reconstruction is probably dodgy." << std::endl;
+      }
+    }
+
+    // Loop over the clusters in the plane and get the hits
+    for(auto const& cluster: clusters){
+          
+      // Get the hits
+      std::vector<art::Ptr<recob::Hit> > hits = fmhc.at(cluster.key());
+      // Get the plane.
+      const geo::PlaneID::PlaneID_t plane(cluster->Plane().Plane);
+       
+      planeHits[plane].insert(planeHits[plane].end(),hits.begin(),hits.end());
+    }
+
+    // Loop over the spacepoints and get the hits
+    sp_hits_tuple.clear();
+    for(auto const& spacepoint : spacepoints){
+      //Get the hits
+      std::vector<art::Ptr<recob::Hit> > hits = fmhsp.at(spacepoint.key());
+
+      // get local Efield at SP_pos
+      TVector3 sp_pos = IShowerTool::GetLArPandoraShowerAlg().SpacePointPosition(spacepoint); 
+      double localEfield_SP = IShowerTool::GetLArPandoraShowerAlg().SCECorrectEField(nominal_Efield, sp_pos);
+
+      // Store the localEfield of all the hits with associated SP's so we can compare with the hits from the clusters later
+      for(auto const& h : hits){
+        sp_hits_tuple.push_back(std::make_tuple(h.key(), localEfield_SP));
+      }           
+    }
+
+    // Calculate the energy for each plane && best plane
+    geo::PlaneID::PlaneID_t bestPlane  = std::numeric_limits<geo::PlaneID::PlaneID_t>::max();
+    unsigned int bestPlaneNumHits = 0;
+
+    //Holder for the final product
+    std::vector<double> ShowerESTAREnergy(fGeom->Nplanes(), -999);
+    std::vector<double> ShowerESTAREnergyError(fGeom->Nplanes(), -999);
+
+    for(auto const& [plane, hits] : planeHits){
+
+      unsigned int planeNumHits = hits.size();
+        
+      //Calculate the Energy
+      double Energy = CalculateEnergy(clockData, detProp, hits, plane);
+
+      if(Energy > 0){    
+        ShowerESTAREnergy.at(plane) = Energy;
+      }
```

Comment:
you may want to use std::numeric_limits to check the tolerance is actually greater than 0